### ISSUE TITLE
uui-swatch: checkmark has too low contrast on light colors

### ISSUE THREAD

**Issue by bjarnef** [umbraco/Umbraco.UI#653]
### Which *exact* UI Library version are you using? For example: 1.0.1 - don't just write v1.

1.6.0-rc

### Bug summary

It seems https://github.com/umbraco/Umbraco.UI/pull/649 changed the `uui-swatch` so checkmark doesn't has sufficient contrast on light background colors, e.g. on yellow background the checkmark used to be black: https://uui.umbraco.com/?path=/story/uui-color-swatches--overview

![image](https://github.com/umbraco/Umbraco.UI/assets/2919859/9378c259-993c-4e1f-9339-051a9c4071d5)

![image](https://github.com/umbraco/Umbraco.UI/assets/2919859/aca8ccce-5012-4e77-9af2-daf9802466a2)

That's why it had the `isLight()` function and set the `color-swatch--light` CSS class.

### Steps to reproduce

Select light colors in `uui-swatches` or `uui-swatch` and notice checkmark is always white.

**Comment by markadrake** [comment]
<img width="494" alt="image" src="https://github.com/umbraco/Umbraco.UI/assets/713355/68a9adc9-0ff4-4e7b-a399-5e887f91d650">

@bjarnef I believe this issue can be closed.

**Comment by bjarnef** [comment]
While it has changes for yellow, it seems to have low contrast for other colors.

![image](https://github.com/umbraco/Umbraco.UI/assets/2919859/2f0fd517-2f82-42d3-a64f-86d71a2c24f9)

which should show this instead:

![image](https://github.com/umbraco/Umbraco.UI/assets/2919859/25caa336-ced1-40d9-a038-b4bc2cdb0b55)

**Comment by markadrake** [comment]
@bjarnef what page of the UUI site are you using for your screenshot? 

Brown looks good:

![image](https://github.com/umbraco/Umbraco.UI/assets/713355/32605693-044f-4aa1-9307-9071c751abaa)

There are some questionable ones, for example:

![image](https://github.com/umbraco/Umbraco.UI/assets/713355/d49674b8-047c-4e8c-9723-aaf4e9727d71)

Perhaps overlaying a check isn't the right, or "most accessible" solution anyways.

**Comment by nielslyngsoe** [comment]
Hi @bjarnef and @markadrake 

Thanks for elaborating on this.

I think this could be fixed by aligning the selection style with other components.

This would mean instead of using the checkmark, we should show the selection by appending a 2px think border with the UUI selection color. To avoid color collision a transparent space (2px?) between the two would be needed. So to avoid it becoming too big the selection state would most like shrink the actual color box.

Here is a somehow similar approach to conceptualize the idea:
<img width="152" alt="image" src="https://github.com/user-attachments/assets/be24cd36-5e89-45fe-9f12-b540d395dba1">

Feel free to bring a proposal for this, thanks

**Comment by bjarnef** [comment]
@nielslyngsoe yes, outline/border highlight be an option.

![image](https://github.com/user-attachments/assets/e463e573-ecf0-4f8d-b8ba-93e9696a2df4)

I think the is light/dark approach could be useful in some scenarios based on the selection style. Or perhaps for other features.

![image](https://github.com/user-attachments/assets/3e9efad1-1ec4-4423-b0b3-98f56916dc07)

**Comment by nielslyngsoe** [comment]
Hi @bjarnef 

sounds good, and good reference you have there.
I would not mind making the color option round, similar to your screenshot here:
![image](https://github.com/user-attachments/assets/eda5f9d3-1489-4a7a-8a23-7bdd1f56067d)

And here was another reference i just stumbled upon:
![image](https://github.com/user-attachments/assets/6c77c7a8-8e8b-439d-97ef-3ead7d158da5)

**Comment by OwainWilliams** [comment]
Hi everyone, 
I was looking to make a PR to add more colours to the colour picker (to match U13) for document type icons and found this chat. 

I'd be happy to give it a go and see if I can make the icons circles with borders on them if it's something that would be accepted. 

@nielslyngsoe

**Comment by bjarnef** [comment]
Another issue is when we have transparent swatches.

![Image](https://github.com/user-attachments/assets/45de5229-ab9f-436e-982b-14d8ef0b5e16)

A solution may be to have a border besides the outline (which may collide with swatch color itself).
https://react.fluentui.dev/?path=/docs/components-swatchpicker--docs

**Comment by bjarnef** [comment]
I think it could work with a gap between border/outline. Without label I think possible a simple tooltip component would be useful: https://github.com/umbraco/Umbraco.UI/discussions/1082 (we have a popover, but I think the library is missing a basic tooltip component).

Some inspiration:

![Image](https://github.com/user-attachments/assets/fb5221dd-5d2e-4a68-945d-ad18dbf0b297)

![Image](https://github.com/user-attachments/assets/2405b157-b755-445d-a156-6a5f0e1400a6)

![Image](https://github.com/user-attachments/assets/2f50e858-cd32-467d-8852-9e9eab2f9221)


As swatch also recently was change from using `background-color` to `background` supporting gradient colors, but it could also be useful for these use-cases, e.g. two-tone color and pattern/image.

![Image](https://github.com/user-attachments/assets/5ba8d318-c868-4a34-859f-cb5ec90ac8c8)

**Comment by bjarnef** [comment]
In a current project based on Umbraco 16.2.0

Not easy to see selection for the blue `#1e91e6` and orange  `#fa6419` colors :)

<img width="377" height="51" alt="Image" src="https://github.com/user-attachments/assets/ced38ccd-e5f9-46d0-acf1-86368c82dce3" />

<img width="367" height="58" alt="Image" src="https://github.com/user-attachments/assets/cf859b25-8324-435d-8e16-5ae2358f14d2" />

<img width="377" height="57" alt="Image" src="https://github.com/user-attachments/assets/22a15ed2-16a6-4506-b4ba-34bc2dd783f4" />

<img width="380" height="63" alt="Image" src="https://github.com/user-attachments/assets/5d5a86f1-4c53-477f-88ac-debb3c93badf" />

I used this internally in a custom view for same colors in background sections:
https://decipher.dev/30-seconds-of-typescript/docs/hexToRGB/

```
 protected override firstUpdated(changedProperties: PropertyValueMap<any> | Map<PropertyKey, unknown>): void {
     super.firstUpdated(changedProperties);

     const bgColor = this.settings?.backgroundColor ? (this.settings?.backgroundColor as any)?.value : '#fff';

     this.style.setProperty('--section-bg-color', bgColor);
     this.style.setProperty('--section-text-color', this.#contrast(bgColor) === 'dark' ? '#fff' : '#000');
 }

 #contrast(hex: string): string {
     const rgb = this.#hexToRgb(hex);
     const o = Math.round(((parseInt(rgb[0]) * 299) + (parseInt(rgb[1]) * 587) + (parseInt(rgb[2]) * 114)) / 1000);

     return (o <= 180) ? 'dark' : 'light';
 };

 // Hex to RGB https://decipher.dev/30-seconds-of-typescript/docs/hexToRGB/
 #hexToRgb(hex: string): string {
     hex = hex.startsWith("#") ? hex.slice(1) : hex;
     if (hex.length === 3) {
         hex = Array.from(hex).reduce((str, x) => str + x + x, ""); // 123 -> 112233
     }

     const values = hex
         .split(/([a-z0-9]{2,2})/)
         .filter(Boolean)
         .map((x) => parseInt(x, 16));

     return `rgb${values.length == 4 ? "a" : ""}(${values.join(", ")})`;
 }; 
```
